### DISCUSSION TITLE
MSGPACK_DEFINE_ARRAY equivalent?

### DISCUSSION THREAD

**Discussion by qiangxinglin** [nlohmann/json#4147] [Q&A]
IMO
`MSGPACK_DEFINE_MAP == NLOHMANN_DEFINE_TYPE_INTRUSIVE`, which would both add some boilerplate code into the class definition to make the custom type conversion. E.g.
```
struct person {
    std::string name;
    std::string address;
    int age;
};
<=>
{"name":"Ned Flanders", "address":"744 Evergreen Terrace", "age":60}
```
Is there any equivalent for `MSGPACK_DEFINE_ARRAY`?
```
struct Bounding_Box
{
    float x1, y1, x2, y2;
    MSGPACK_DEFINE_ARRAY(x1, y1, x2, y2)
};
<=>
[1.0, 2.0, 3.0, 4.0]
```

**Comment by nlohmann** [comment]
No, there is no macro equivalent to MSGPACK_DEFINE_ARRAY. All of the NLOHMANN_DEFINE_TYPE_INTRUSIVE/NON_INTRUSIVE (and _WITH_DEFAULT, _WITH_NAMES) macros expand to nlohmann_json_j[#v1] = nlohmann_json_t.v1 for each member (see include/nlohmann/detail/macro_scope.hpp), which always produces a named-key JSON object, never a positional array. To get array-style output for Bounding_Box, you'd write to_json/from_json by hand, building/reading a json::array() of the four floats in order.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
